In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


# **STEP 1: Read and view file**

In [ ]:
spams = pd.read_csv(r"/content/drive/My Drive/0. STUDYING Peo Peo/PYTHON/spam_original.csv")
spams.head()

word_freq_make   word_freq_address  ...   capital_run_length_total   spam
0            0.00                0.64  ...                        278      1
1            0.21                0.28  ...                       1028      1
2            0.06                0.00  ...                       2259      1
3            0.00                0.00  ...                        191      1
4            0.00                0.00  ...                        191      1

[5 rows x 58 columns]

# **STEP 2: Kiểm tra dữ liệu lỗi**

In [ ]:
spams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   word_freq_make               4601 non-null   float64
 1    word_freq_address           4601 non-null   float64
 2    word_freq_all               4601 non-null   float64
 3    word_freq_3d                4601 non-null   float64
 4    word_freq_our               4601 non-null   float64
 5    word_freq_over              4601 non-null   float64
 6    word_freq_remove            4601 non-null   float64
 7    word_freq_internet          4601 non-null   float64
 8    word_freq_order             4601 non-null   float64
 9    word_freq_mail              4601 non-null   float64
 10   word_freq_receive           4601 non-null   float64
 11   word_freq_will              4601 non-null   float64
 12   word_freq_people            4601 non-null   float64
 13   word_freq_report 

# **STEP 3**: 
##**1. Tách X va y (dataframe)**







In [ ]:
X = spams.iloc[:, :-1]               ## X là dữ liệu bỏ đi cột cuối cùng Outcome và lấy tất cả các dòng trong table
y = spams.iloc[:, -1].to_numpy()     ## y thanh array

print(y[:5])
X.head()

[1 1 1 1 1]


word_freq_make  ...   capital_run_length_total
0            0.00  ...                        278
1            0.21  ...                       1028
2            0.06  ...                       2259
3            0.00  ...                        191
4            0.00  ...                        191

[5 rows x 57 columns]

##**2. Train_Test**


In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, 
                                                    stratify = y, random_state = 0)

#**STEP 4: Choose model**

##**1. Model: Support Vector Machine (SVM)**

In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf_svm = SVC(kernel = 'poly', C = 10)   
clf_svm.fit(X_train, y_train)

print("Support Vector Machine")
print("\ntrain acc: ", accuracy_score(y_train, clf_svm.predict(X_train)))
print("train acc: ", accuracy_score(y_test, clf_svm.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf_svm.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf_svm.predict(X_test), digits = 5))

Support Vector Machine

train acc:  0.6720496894409937
train acc:  0.6618392469225199

Confusion matrix : 
 [[828   9]
 [458  86]]
Classification report : 
               precision    recall  f1-score   support

           0    0.64386   0.98925   0.78003       837
           1    0.90526   0.15809   0.26917       544

    accuracy                        0.66184      1381
   macro avg    0.77456   0.57367   0.52460      1381
weighted avg    0.74683   0.66184   0.57879      1381



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 1381 dữ liệu của test-set thì có:

*   ####  Trong y_test: 
   - spam = 0 : 
      - Thực tế: 837
      - Dự đoán chính xác:  828
      - Dự đoán sai: 9
    - spam = 1 : 
      - Thực tế: 544
      - Dự đoán chính xác:  458
      - Dự đoán sai: 86

* #### Bài toán: accuracy  = 66,184%; 
==> Chủ yếu dự đoán sai nhóm spam = 1. 


##**2. Model: Logistic Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf = LogisticRegression(max_iter = len(y_train))
clf.fit(X_train,y_train)

print("Logistic Regression")
print("\ntrain acc: ", accuracy_score(y_train, clf.predict(X_train)))
print("train acc: ", accuracy_score(y_test, clf.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf.predict(X_test), digits = 5))

Logistic Regression

train acc:  0.9316770186335404
train acc:  0.9290369297610427

Confusion matrix : 
 [[792  45]
 [ 53 491]]
Classification report : 
               precision    recall  f1-score   support

           0    0.93728   0.94624   0.94174       837
           1    0.91604   0.90257   0.90926       544

    accuracy                        0.92904      1381
   macro avg    0.92666   0.92441   0.92550      1381
weighted avg    0.92891   0.92904   0.92894      1381



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 1381 dữ liệu của test-set thì có:

*   ####  Trong y_test: 
   - spam = 0 : 
      - Thực tế: 837
      - Dự đoán chính xác: 792
      - Dự đoán sai: 45
    - spam = 1 : 
      - Thực tế: 544
      - Dự đoán chính xác:  53
      - Dự đoán sai: 491

* #### Bài toán: accuracy  = 92,904%; 
==> Chủ yếu dự đoán sai nhóm spam = 1. 


#**3. Model: Decision Tree Classifier**

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf = DecisionTreeClassifier()
clf.fit(X_train,y_train)

print("Decision Tree Classifier")
print("\ntrain acc: ", accuracy_score(y_train, clf.predict(X_train)))
print("train acc: ", accuracy_score(y_test, clf.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf.predict(X_test), digits = 5))

Decision Tree Classifier

train acc:  0.9993788819875776
train acc:  0.9225199131064447

Confusion matrix : 
 [[781  56]
 [ 51 493]]
Classification report : 
               precision    recall  f1-score   support

           0    0.93870   0.93309   0.93589       837
           1    0.89800   0.90625   0.90210       544

    accuracy                        0.92252      1381
   macro avg    0.91835   0.91967   0.91900      1381
weighted avg    0.92267   0.92252   0.92258      1381



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 1381 dữ liệu của test-set thì có:

*   ####  Trong y_test: 
   - spam = 0 : 
      - Thực tế: 837
      - Dự đoán chính xác: 781
      - Dự đoán sai: 56
    - spam = 1 : 
      - Thực tế: 544
      - Dự đoán chính xác:  51
      - Dự đoán sai: 493

* #### Bài toán: accuracy  = 92,252%; 
==> Chủ yếu dự đoán sai nhóm spam = 1. 

#**4. Model: Random Forest Classifier**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf = RandomForestClassifier()
clf.fit(X_train,y_train)

print("Random Forest Classifier")
print("\ntrain acc: ", accuracy_score(y_train, clf.predict(X_train)))
print("train acc: ", accuracy_score(y_test, clf.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf.predict(X_test), digits = 5))

Random Forest Classifier

train acc:  0.9993788819875776
train acc:  0.9580014482259233

Confusion matrix : 
 [[811  26]
 [ 32 512]]
Classification report : 
               precision    recall  f1-score   support

           0    0.96204   0.96894   0.96548       837
           1    0.95167   0.94118   0.94640       544

    accuracy                        0.95800      1381
   macro avg    0.95686   0.95506   0.95594      1381
weighted avg    0.95796   0.95800   0.95796      1381



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 1381 dữ liệu của test-set thì có:

*   ####  Trong y_test: 
   - spam = 0 : 
      - Thực tế: 837
      - Dự đoán chính xác: 811
      - Dự đoán sai: 26
    - spam = 1 : 
      - Thực tế: 544
      - Dự đoán chính xác:  32
      - Dự đoán sai: 512

* #### Bài toán: accuracy  = 95,8%; 
==> Chủ yếu dự đoán sai nhóm spam = 1. 